docs/mocksupport.rst

```diff
@@ -867,13 +738,338 @@ This adds the instances to the repository display of the previous example:
+    class MyUserProvider(MethodProvider):
+
+Within that class, the following components must be defined.
+
+1. Constructor (`__init__()`) that takes as input at least the ``cimrepository``
+   object parameter and passes it to the superclass and any other constructor
+   parameters the user-defined provider requires.  Since the user creates the
+   instance.
+
+.. code-block:: python
+
+    def __init__(self, cimrepository):
+        """
+        Init of test provider
+        """
+        super(UserMethodTestProvider, self).__init__(cimrepository)
+
+
+2. Definition of the CIM class the provider supports as a class variable.
+
+.. code-block:: python
+
+    provider_classnames = 'CIM_Foo'
+
+3. Definition and implementation of the WBEM operations supported by the
+   provider from the set of request operations defined in the table above for a
+   specific provider type. This must be a subset of the methods defined for the
+   provider type. Methods that do not exist in the user-defined provider
+   default to the default method. Each of these methods may:
+
+   * provide parameter or CIM repository validation in addition to the normal
+     request validation,
+
+   * modify parameters of the request.
+
+   * generate an CIMError exception for the operation.
+
+   * make modifications to the CIM repository.
+
+   The request may be completed by the user provider method calling the
+   superclass corresponding method to complete its final validation and
+   modify the repository or by the provider directly modifying the repository
+   based on the method and parameters.
+
+    The user-defined providers have access to:
+       * methods defined in their superclass
+       * methods defined in :class:~pywbem_mock.BaseProvider`
+       * methods to access the CIM repository using the methods defined in
+         :class:`~pywbem_mock.InMemoryRepository`
+
+   TODO define the aprori validation for each operation. Good place for table
+
+.. code-block:: python
+
+    # define CreateInstance and call the superclass method
+    def CreateInstance(self, namespace, NewInstance):
+        """Test Create instance just calls super class method"""
+
+        # Simply calls the default responder method for CreateInstance
+        return super(UserInstanceTestProvider, self).CreateInstance(
+            namespace, NewInstance)
+
+    TODO show example where it specifically modifies the repository.
+
+3. Optional method (``post_register-setup()``) that
+   :meth:`~pywbem_mock.FakedWBEMConnection.register_provider` will
+   call as after the provider registration is successful.  This allows the
+   provider to do any special setup it desires uses its own methods. This
+   method includes the current connection as a parameter so that client
+   methods can be executed directly.
+
+.. code-block:: python
+
+    def post_register_setup(self, conn):
+        ....
+
+
+.. _`Creating instance user-defined providers`:
+
+Creating instance user-defined providers
```

Comment:
Change to "instance user-defined providers" to "user-defined instance write providers"?

Occurs multiple times.

Reply:
DONE